src/main/scala/firrtl/transforms/FlattenRegUpdate.scala

```diff
@@ -58,38 +58,80 @@ object FlattenRegUpdate {
     * @return [[firrtl.ir.Module Module]] with register updates flattened
     */
   def flattenReg(mod: Module): Module = {
-    // We want to flatten Mux trees for reg updates into if-trees for
-    // improved QoR for conditional updates.  However, unbounded recursion
-    // would take exponential time, so don't redundantly flatten the same
-    // Mux more than a bounded number of times, preserving linear runtime.
-    // The threshold is empirical but ample.
-    val flattenThreshold = 4
-    val numTimesFlattened = mutable.HashMap[Mux, Int]()
-    def canFlatten(m: Mux): Boolean = {
-      val n = numTimesFlattened.getOrElse(m, 0)
-      numTimesFlattened(m) = n + 1
-      n < flattenThreshold
-    }
+    // We want to flatten Mux trees for reg updates into if-trees for improved QoR for conditional
+    // updates.  Sometimes the fan-in for a register has a mux structure with repeated
+    // sub-expressions that are themselves complex mux structures. These repeated structures can
+    // cause explosions in the size and complexity of the Verilog. In addition, user code that
+    // follows such structure often will have conditions in the sub-trees that are mutually
+    // exclusive with the conditions in the muxes closer to the register input. For example:
+    //
+    // when a :      ; when 1
+    //   r <= foo
+    // when b :      ; when 2
+    //   when a :
+    //     r <= bar  ; when 3
+    //
+    // After expand whens, when 1 is a common sub-expression that will show up twice in the mux
+    // structure from when 2:
+    //
+    // _GEN_0 = mux(a, foo, r)
+    // _GEN_1 = mux(a, bar, _GEN_0)
+    // r <= mux(b, _GEN_1, _GEN_0)
+    //
+    // Inlining _GEN_0 into _GEN_1 would result in unreachable lines in the Verilog. While we could
+    // do some optimizations here, this is *not* really a problem, it's just that Verilog metrics
+    // are based on the assumption of human-written code and as such it results in unreachable
+    // lines. Simply not inlining avoids this issue and leaves the optimizations up to synthesis
+    // tools which do a great job here.
```

Comment:
I like this explanation!

Is this related to the changes in https://github.com/freechipsproject/firrtl/pull/804 ?

Reply:
They are indeed related. #804 was an earlier attempt to solve the same problem. I believe that this approach solves the actual issue better (unreachable branches), but #804 is useful for other reasons. It's a more compiler-y optimization attempt whereas this solution is more recognizing that the existing inlining behavior is actually a de-optimization (ie. undoing Common SubExpression Elimination). The fact that this approach works so well is more of a property of last connect semantics than something fundamental, but it's still as useful observation and heuristic.

I think #804 is still worth investigating, since the Verilog from the original example is still kind of ugly:
```verilog
  always @(posedge clock) begin
    if (reset) begin
      reg_ <= 8'h0;
    end else if (io_b) begin
      if (io_c) begin
        if (io_d) begin
          reg_ <= 8'h2;
        end else begin
          reg_ <= _GEN_1;
        end
      end else begin
        reg_ <= _GEN_1;
      end
    end else begin
      reg_ <= _GEN_1;
    end
  end
```
#804 would turn this into
```verilog
  always @(posedge clock) begin
    if (reset) begin
      reg_ <= 8'h0;
    end else if (io_b && io_c && io_d) begin
      reg_ <= 8'h2;
    end else begin
      reg_ <= _GEN_1;
    end
  end
```